test/framework/src/main/java/org/elasticsearch/cluster/routing/RoutingNodesHelper.java

```diff
@@ -12,41 +12,29 @@
 
 import java.util.ArrayList;
 import java.util.List;
+import java.util.Objects;
 import java.util.stream.IntStream;
 import java.util.stream.Stream;
 
+import static java.util.stream.Collectors.toCollection;
+import static org.elasticsearch.common.util.CollectionUtils.iterableAsArrayList;
+
 public final class RoutingNodesHelper {
 
     private RoutingNodesHelper() {}
 
     public static List<ShardRouting> shardsWithState(RoutingNodes routingNodes, ShardRoutingState state) {
-        List<ShardRouting> shards = new ArrayList<>();
-        if (state == ShardRoutingState.UNASSIGNED) {
-            routingNodes.unassigned().forEach(shards::add);
-        } else {
-            for (RoutingNode routingNode : routingNodes) {
-                shards.addAll(routingNode.shardsWithState(state));
-            }
-        }
-        return shards;
+        return state == ShardRoutingState.UNASSIGNED
+            ? iterableAsArrayList(routingNodes.unassigned())
+            : routingNodes.stream()
+                .flatMap(routingNode -> routingNode.shardsWithState(state).stream())
+                .collect(toCollection(ArrayList::new));
```

Comment:
Some tests modify the output of this method (eg shuffle)

Reply:
I think I'd rather have those tests copy the result into a new list instead and make this immutable. It's only a couple of spots to change I think.